core/src/main/scala/kafka/log/LogManager.scala

```diff
@@ -56,7 +60,9 @@ class LogManager(val logDirs: Array[File],
                  scheduler: Scheduler,
                  val brokerState: BrokerState,
                  brokerTopicStats: BrokerTopicStats,
-                 time: Time) extends Logging {
+                 val zkUtils: ZkUtils,
```

Comment:
@lindong28 : Good point on the circular dependency on constructing those objects. Another approach that I am thinking is to have a DiskFailureChannel which contains an in-memory queue. DiskFailureChannel can be passed into all components such as LogMananger, LogCleaner, ReplicaManager, etc where IO errors can be generated. If an IO error happens in a component, it just enqueues the disk name into the DiskFailureChannel. We can have a separate thread (maybe in ReplicaManager) that reads from DiskFailureChannel and acts on it (e.g., remove partitions in ReplicaManager, remove logs in LogManager, remove partitions from replica fetcher), etc. This way, all disk errors will be handled consistently. The only thing is that disk errors will be handled asynchronously. However, I am not sure if disk errors need to be processed synchronously.